File: packages/runar-py/runar/sdk/script_vm.py
"""Bitcoin Script virtual machine for off-chain testing and debugging.

Thin wrapper around the bsv-blockchain ``bsv-sdk`` ``Spend`` script
interpreter (``bsv.script.spend.Spend``). It does NOT re-implement Bitcoin
Script — the upstream interpreter does all execution. The wrapper drives
``Spend.step()`` and records a per-opcode trace so it can expose the same
step-mode debugger API as the TypeScript and Go ScriptVMs::

    vm = ScriptVM()
    res = vm.execute_hex("5253935587")          # one-shot execution
    vm.load_hex("", "5253935587")               # step-mode: load...
    while (step := vm.step()) is not None: ...   # ...then walk opcodes

``bsv-sdk`` is an **optional** dependency — ``runar-py`` itself has zero
required dependencies. Install the ScriptVM extra to use this module::

    pip install runar[script-vm]

If ``bsv-sdk`` is not installed, importing this module still succeeds; the
``ImportError`` is raised only when a ``ScriptVM`` method is actually
called, with a message pointing at the extra.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

_BSV_IMPORT_HINT = (
    "runar.sdk.script_vm requires the optional 'bsv-sdk' dependency. "
    "Install it with:  pip install runar[script-vm]  (or: pip install bsv-sdk)"
)


def _require_bsv():
    """Lazily import bsv-sdk, raising a helpful error if it is absent."""
    try:
        from bsv.script.script import Script
        from bsv.script.spend import Spend
        from bsv.constants import OPCODE_VALUE_NAME_DICT
    except ImportError as exc:  # pragma: no cover - exercised only without the extra
        raise ImportError(_BSV_IMPORT_HINT) from exc
    return Script, Spend, OPCODE_VALUE_NAME_DICT


@dataclass
class VMOptions:
    """Configuration for a :class:`ScriptVM`. The default is a usable VM."""

    # Reserved for future flags; kept so the constructor signature is stable
    # across tiers. bsv-sdk's Spend has no per-instance flag surface.
    pass


@dataclass
class VMResult:
    """The outcome of a full script execution."""

    success: bool
    stack: list[bytes] = field(default_factory=list)
    alt_stack: list[bytes] = field(default_factory=list)
    error: Optional[str] = None
    ops_executed: int = 0


@dataclass
class StepResult:
    """The outcome of executing a single opcode in step mode."""

    offset: int
    opcode: str
    main_stack: list[bytes] = field(default_factory=list)
    alt_stack: list[bytes] = field(default_factory=list)
    error: Optional[str] = None
    context: str = ""  # "unlocking" or "locking"


def _opcode_name(op: bytes, names: dict) -> str:
    """Resolve an opcode byte to a name. Direct-push chunks (0x01..0x4b)
    are not in the name table — report them as ``PUSH_<n>``."""
    name = names.get(op)
    if name is not None:
        return name
    if len(op) == 1 and 0x01 <= op[0] <= 0x4B:
        return f"PUSH_{op[0]}"
    return f"OP_UNKNOWN_0x{op.hex()}"


class ScriptVM:
    """Executes Bitcoin Script bytes via the bsv-sdk ``Spend`` interpreter
    and records a step trace for debugger-style inspection."""

    def __init__(self, options: Optional[VMOptions] = None) -> None:
        self._options = options or VMOptions()
        self._trace: list[StepResult] = []
        self._result: Optional[VMResult] = None
        self._pc = 0
        self._loaded = False

    # ------------------------------------------------------------------
    # One-shot execution
    # ------------------------------------------------------------------

    def execute(self, unlocking: bytes, locking: bytes) -> VMResult:
        """Run the unlocking script followed by the locking script and
        return the final VM state. An empty unlocking script is allowed."""
        trace, result = self._run(unlocking, locking)
        self._trace = trace
        self._result = result
        self._pc = 0
        self._loaded = True
        return result

    def execute_hex(self, script_hex: str) -> VMResult:
        """Run a single hex-encoded script (as the locking script, with an
        empty unlocking script) and return the final VM state."""
        return self.execute(b"", bytes.fromhex(script_hex))

    # ------------------------------------------------------------------
    # Step mode (debugger API)
    # ------------------------------------------------------------------

    def load(self, unlocking: bytes, locking: bytes) -> None:
        """Prepare the VM to step through the unlocking + locking scripts.
        The interpreter runs the scripts immediately; :meth:`step` then walks
        the recorded trace one opcode at a time."""
        trace, result = self._run(unlocking, locking)
        self._trace = trace
        self._result = result
        self._pc = 0
        self._loaded = True

    def load_hex(self, unlocking_hex: str, locking_hex: str) -> None:
        """Prepare the VM to step through hex-encoded unlocking + locking
        scripts."""
        self.load(bytes.fromhex(unlocking_hex), bytes.fromhex(locking_hex))

    def step(self) -> Optional[StepResult]:
        """Execute the next opcode and return its :class:`StepResult`, or
        ``None`` when the script has finished (or nothing is loaded)."""
        if not self._loaded or self._pc >= len(self._trace):
            return None
        step = self._trace[self._pc]
        self._pc += 1
        return step

    @property
    def pc(self) -> int:
        """The current step cursor (opcodes stepped so far)."""
        return self._pc

    @property
    def is_complete(self) -> bool:
        """Whether stepping has reached the end of the trace."""
        return self._loaded and self._pc >= len(self._trace)

    @property
    def is_success(self) -> bool:
        """Whether the loaded script executed successfully."""
        return self._result is not None and self._result.success

    @property
    def current_stack(self) -> list[bytes]:
        """The main stack after the last stepped opcode (or the final stack
        once stepping is complete / after a one-shot ``execute``)."""
        if 0 < self._pc <= len(self._trace):
            return self._trace[self._pc - 1].main_stack
        if self._result is not None:
            return self._result.stack
        return []

    @property
    def current_alt_stack(self) -> list[bytes]:
        """The alt stack at the current cursor, mirroring ``current_stack``."""
        if 0 < self._pc <= len(self._trace):
            return self._trace[self._pc - 1].alt_stack
        if self._result is not None:
            return self._result.alt_stack
        return []

    @property
    def context(self) -> str:
        """``"unlocking"`` or ``"locking"`` — which script the cursor is in."""
        if 0 < self._pc <= len(self._trace):
            return self._trace[self._pc - 1].context
        return ""

    # ------------------------------------------------------------------
    # internals
    # ------------------------------------------------------------------

    def _run(self, unlocking: bytes, locking: bytes) -> tuple[list[StepResult], VMResult]:
        Script, Spend, names = _require_bsv()

        spend = Spend({
            "sourceTXID": "00" * 32,
            "sourceOutputIndex": 0,
            "sourceSatoshis": 0,
            "lockingScript": Script(locking),
            "transactionVersion": 1,
            "otherInputs": [],
            "outputs": [],
            "inputIndex": 0,
            "unlockingScript": Script(unlocking),
            "inputSequence": 0xFFFFFFFF,
            "lockTime": 0,
        })

        trace: list[StepResult] = []
        error: Optional[str] = None

        def _ctx_label() -> str:
            return "unlocking" if spend.context == "UnlockingScript" else "locking"

        def _at_end() -> bool:
            return (
                spend.context == "LockingScript"
                and spend.program_counter >= len(spend.locking_script.chunks)
            )

        # Drive Spend.step() one opcode at a time, snapshotting state after
        # each. Spend.step() auto-transitions UnlockingScript -> LockingScript.
        try:
            while not _at_end():
                # Resolve the chunk about to run (handle the auto-transition
                # Spend.step() performs internally at the unlocking-script end).
                if (
                    spend.context == "UnlockingScript"
                    and spend.program_counter >= len(spend.unlocking_script.chunks)
                ):
                    spend.context = "LockingScript"
                    spend.program_counter = 0
                    continue
                if spend.context == "UnlockingScript":
                    chunk = spend.unlocking_script.chunks[spend.program_counter]
                else:
                    chunk = spend.locking_script.chunks[spend.program_counter]

                offset = spend.program_counter
                ctx_label = _ctx_label()
                spend.step()
                trace.append(StepResult(
                    offset=offset,
                    opcode=_opcode_name(chunk.op, names),
                    main_stack=list(spend.stack),
                    alt_stack=list(spend.alt_stack),
                    context=ctx_label,
                ))
        except Exception as exc:  # noqa: BLE001 - upstream raises bare Exception
            error = str(exc)
            if trace:
                trace[-1].error = error

        # Determine success. bsv-sdk's Spend.validate() raises on any
        # failure; here we replicate its final checks without re-running.
        success = False
        if error is None:
            balanced = len(spend.if_stack) == 0
            non_empty = len(spend.stack) > 0
            if balanced and non_empty:
                success = Spend.cast_to_bool(spend.stack[-1])
            if not success and error is None:
                # Mirror Spend.validate()'s diagnostic for a clean falsy result.
                error = "script evaluated to false (top stack element not truthy)"

        result = VMResult(
            success=success,
            stack=list(spend.stack),
            alt_stack=list(spend.alt_stack),
            error=error if not success else None,
            ops_executed=len(trace),
        )
        return trace, result

File: packages/runar-testing/src/vm/script-vm.ts
/**
 * Bitcoin Script Virtual Machine for BSV.
 *
 * This is NOT a custom VM: the execution core is the upstream `@bsv/sdk`
 * `Spend` interpreter — the same production engine that validates real BSV
 * transactions. `ScriptVM` only
 *
 *   1. builds a synthetic single-input transaction context so bare scripts
 *      (with no real tx) can be run,
 *   2. drives `Spend.step()` one opcode at a time — the TS equivalent of the
 *      Go tier's `go-sdk/script/interpreter` `Debugger` hook — so the stack can
 *      be observed per opcode (step-mode debugger, used by `runar-cli debug`),
 *   3. applies the harness-level DoS bounds (`maxOps`, `maxStackSize`,
 *      `maxScriptSize`, `InputLimits.MAX_SCRIPT_BYTES`), and
 *   4. shapes the result into {@link VMResult}.
 *
 * Every opcode's semantics — arithmetic, byte-string ops, OP_LSHIFT/OP_RSHIFT
 * truncation, OP_NUM2BIN sign handling, OP_RETURN termination, real ECDSA for
 * OP_CHECKSIG / OP_CHECKMULTISIG — comes from `Spend`. Nothing here reimplements
 * an opcode. This matters because `ScriptVM` is the script-semantics half of the
 * source-vs-script differential oracle (`oracle/differential-execution.ts`): an
 * oracle written by the same project it grades is not independent.
 *
 * ## Signatures are REAL — there is no mock checksig
 *
 * The previous hand-rolled VM took a `checkSigCallback` that defaulted to
 * `() => true`, i.e. every `OP_CHECKSIG` passed. That is fail-open: a contract
 * whose only guard is a signature check "verified" against nothing. `Spend`
 * performs real secp256k1 verification against the BIP-143 sighash of the
 * synthetic context below, so a script containing `OP_CHECKSIG` only succeeds
 * when the caller supplies a signature that is genuinely valid for that context
 * (see `oracle/real-crypto-execution.ts`, which builds exactly such witnesses).
 * `checkSigCallback` is gone; there is no hand-written signature path.
 *
 * ## Accept rule and the consensus wrappers
 *
 * `ScriptVM`'s contract is "execute these bytes and report the final stack", so
 * `success` means *no evaluation error and a truthy top-of-stack* — the same
 * rule the previous VM used, and the rule that lets `executeHex()` run a bare
 * opcode fragment that legitimately leaves several items on the stack.
 *
 * The three consensus rules that `Spend.validate()` layers ON TOP of evaluation
 * — push-only unlocking scripts, the clean-stack rule, and minimal-push /
 * low-S encoding — are deliberately NOT applied here, because they are
 * meaningless for a fragment. They are enforced where a whole spend is being
 * judged: `oracle/tri-modal-execution.ts` (strict `Spend.validate()` leg) and
 * `oracle/real-crypto-execution.ts`. Pass `flags.strictEncoding` to turn the
 * encoding rules back on for this VM.
 */

import { LockingScript, UnlockingScript, Spend } from '@bsv/sdk';
import { opcodeName } from './opcodes.js';
import { isTruthy, hexToBytes } from './utils.js';
import { InputLimits, CanonicalJsonError } from 'runar-ir-schema';

// ---------------------------------------------------------------------------
// Public interfaces
// ---------------------------------------------------------------------------

export interface VMResult {
  success: boolean;
  stack: Uint8Array[];
  altStack: Uint8Array[];
  error?: string;
  opsExecuted: number;
  maxStackDepth: number;
}

export interface VMOptions {
  /** Maximum number of non-push opcodes to execute (default 500_000). */
  maxOps?: number;
  /** Maximum number of items on the main + alt stack (default 1_000). */
  maxStackSize?: number;
  /** Maximum script size in bytes (default: unlimited for BSV). */
  maxScriptSize?: number;
  /** Behavioural flags. */
  flags?: VMFlags;
}

export interface VMFlags {
  /**
   * Apply the upstream engine's strict-encoding consensus rules: minimally
   * encoded pushes, minimally encoded script numbers, and low-S signatures.
   * Default `false` (relaxed), matching the permissiveness callers rely on when
   * executing hand-written opcode fragments.
   */
  strictEncoding?: boolean;
}

/** Result of a single step in the VM. */
export interface StepResult {
  /** Byte offset of the opcode that was executed. */
  offset: number;
  /** Name of the opcode (e.g. 'OP_ADD', 'OP_DUP', 'PUSH_20'). */
  opcode: string;
  /** Main stack after this opcode. */
  mainStack: Uint8Array[];
  /** Alt stack after this opcode. */
  altStack: Uint8Array[];
  /** Set if the opcode caused an error (e.g. OP_VERIFY on false). */
  error?: string;
  /** Which script is executing. */
  context: 'unlocking' | 'locking';
}

// ---------------------------------------------------------------------------
// Synthetic transaction context
// ---------------------------------------------------------------------------

/**
 * `Spend` needs a transaction to compute the BIP-143 sighash for OP_CHECKSIG.
 * A bare script has none, so we supply a deterministic single-input, zero-output
 * spend of a null outpoint. Identical in shape to the context used by
 * `oracle/real-crypto-execution.ts`, so a signature built for one validates on
 * the other. `transactionVersion` is 1 so that `Spend`'s relaxation switch stays
 * under our control via `flags.strictEncoding` (it auto-relaxes for version > 1).
 */
const SYNTHETIC_SOURCE_TXID = '00'.repeat(32);
const SYNTHETIC_SOURCE_SATOSHIS = 100_000;
const SYNTHETIC_TX_VERSION = 1;

/**
 * The synthetic spend context, exported so anything that needs to SIGN for this
 * VM derives the BIP-143 preimage from the same values the VM verifies against
 * instead of copying them. The version differs from
 * `oracle/real-crypto-execution.ts` (1 here, 2 there) and version is part of the
 * preimage, so a signature built against the wrong one silently fails to verify
 * — a copy that drifts would look like a codegen defect.
 */
export const SYNTHETIC_SPEND_CONTEXT = {
  sourceTXID: SYNTHETIC_SOURCE_TXID,
  sourceOutputIndex: 0,
  sourceSatoshis: SYNTHETIC_SOURCE_SATOSHIS,
  transactionVersion: SYNTHETIC_TX_VERSION,
  inputIndex: 0,
  inputSequence: 0xffffffff,
  lockTime: 0,
} as const;

/** Opcodes above OP_16 count towards the operation limit (consensus rule). */
const OP_16 = 0x60;
const OP_PUSHDATA1 = 0x4c;
const OP_PUSHDATA2 = 0x4d;
const OP_PUSHDATA4 = 0x4e;
const OP_IF = 0x63;
const OP_NOTIF = 0x64;
const OP_VERIF = 0x65;
const OP_VERNOTIF = 0x66;
const OP_ENDIF = 0x68;
const OP_RETURN = 0x6a;

// ---------------------------------------------------------------------------
// Script execution error (harness-level: limits, not opcode semantics)
// ---------------------------------------------------------------------------

class ScriptError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ScriptError';
  }
}

/**
 * `ScriptEvaluationError.message` embeds a full hex dump of both stacks, which
 * for a 900 KB SLH-DSA script is megabytes of noise. Keep the first line and
 * strip the fixed prefix so the reported error stays readable and still names
 * the offending opcode.
 */
function describeError(e: unknown): string {
  const raw = e instanceof Error ? e.message : String(e);
  const firstLine = raw.split('\n', 1)[0] ?? raw;
  return firstLine.startsWith('Script evaluation error: ')
    ? firstLine.slice('Script evaluation error: '.length)
    : firstLine;
}

// ---------------------------------------------------------------------------
// Chunk byte offsets
// ---------------------------------------------------------------------------

/**
 * Byte offset of every chunk `Script.parseChunks` produces, plus a trailing
 * sentinel equal to the script length (so `offsets[i + 1]` is always defined).
 *
 * `Spend`'s program counter is a CHUNK index; callers of the step API
 * (`runar-cli debug`, source-map lookups) want a BYTE offset. This mirrors
 * `Script.parseChunks` exactly — including its post-Genesis behaviour of
 * swallowing everything after a top-level OP_RETURN into that chunk, and its
 * clamping of pushes that run past the end of the script — so chunk indices
 * line up one-for-one.
 */
function chunkByteOffsets(script: Uint8Array): number[] {
  const offsets: number[] = [];
  const length = script.length;
  let pos = 0;
  let inConditionalBlock = 0;

  while (pos < length) {
    offsets.push(pos);
    const op = script[pos++] ?? 0;

    if (op === OP_RETURN && inConditionalBlock === 0) {
      break; // rest of the script is this chunk's data
    }
    if (op === OP_IF || op === OP_NOTIF || op === OP_VERIF || op === OP_VERNOTIF) {
      inConditionalBlock++;
    } else if (op === OP_ENDIF) {
      inConditionalBlock--;
    }

    if (op > 0 && op < OP_PUSHDATA1) {
      pos = Math.min(pos + op, length);
    } else if (op === OP_PUSHDATA1) {
      const len = pos < length ? (script[pos++] ?? 0) : 0;
      pos = Math.min(pos + len, length);
    } else if (op === OP_PUSHDATA2) {
      const len = (script[pos] ?? 0) | ((script[pos + 1] ?? 0) << 8);
      pos = Math.min(pos + 2, length);
      pos = Math.min(pos + len, length);
    } else if (op === OP_PUSHDATA4) {
      const len =
        (((script[pos] ?? 0) |
          ((script[pos + 1] ?? 0) << 8) |
          ((script[pos + 2] ?? 0) << 16) |
          ((script[pos + 3] ?? 0) << 24)) >>>
        0);
      pos = Math.min(pos + 4, length);
      pos = Math.min(pos + len, length);
    }
  }

  offsets.push(length);
  return offsets;
}

/** Human-readable name for a parsed chunk's opcode. */
function chunkOpcodeName(op: number): string {
  if (op >= 0x01 && op <= 0x4b) return `PUSH_${op}`;
  if (op === OP_PUSHDATA1) return 'OP_PUSHDATA1';
  if (op === OP_PUSHDATA2) return 'OP_PUSHDATA2';
  if (op === OP_PUSHDATA4) return 'OP_PUSHDATA4';
  return opcodeName(op);
}

function toBytes(items: number[][]): Uint8Array[] {
  return items.map((i) => Uint8Array.from(i));
}

// ---------------------------------------------------------------------------
// VM
// ---------------------------------------------------------------------------

interface Position {
  context: 'unlocking' | 'locking';
  chunkIndex: number;
}

export class ScriptVM {
  private readonly maxOps: number;
  private readonly maxStackSize: number;
  private readonly maxScriptSize: number;
  readonly flags: VMFlags;

  /** The upstream engine. `null` before the first execute/load. */
  private spend: Spend | null = null;
  private unlockingOffsets: number[] = [0];
  private lockingOffsets: number[] = [0];
  private unlockingLength = 0;
  private lockingLength = 0;

  private _opsExecuted = 0;
  private _maxStackDepth = 0;
  private _isComplete = false;
  private _isSuccess = false;
  private _stepError: string | undefined;

  constructor(options: VMOptions = {}) {
    this.maxOps = options.maxOps ?? 500_000;
    this.maxStackSize = options.maxStackSize ?? 1_000;
    this.maxScriptSize = options.maxScriptSize ?? Number.MAX_SAFE_INTEGER;
    this.flags = options.flags ?? {};
  }

  // -------------------------------------------------------------------------
  // Public API — one-shot execution
  // -------------------------------------------------------------------------

  /**
   * Execute combined unlocking + locking script.
   *
   * The unlocking script is executed first; its resulting stack is the initial
   * stack for the locking script, and the locking script's outcome determines
   * success. This is exactly `Spend`'s own two-phase evaluation.
   */
  execute(unlockingScript: Uint8Array, lockingScript: Uint8Array): VMResult {
    this.assertScriptBytesUnderLimit(unlockingScript);
    this.assertScriptBytesUnderLimit(lockingScript);
    if (
      unlockingScript.length > this.maxScriptSize ||
      lockingScript.length > this.maxScriptSize
    ) {
      this.loadInternal(new Uint8Array(0), new Uint8Array(0));
      return this.buildResult('Script size exceeds maximum');
    }
    this.loadInternal(unlockingScript, lockingScript);
    return this.runToEnd();
  }

  /**
   * Execute a single script (convenience for testing) — modelled as an empty
   * unlocking script followed by `script` as the locking script.
   */
  executeScript(script: Uint8Array): VMResult {
    return this.execute(new Uint8Array(0), script);
  }

  /**
   * Execute a script from a hex string.
   */
  executeHex(scriptHex: string): VMResult {
    return this.executeScript(hexToBytes(scriptHex));
  }

  /**
   * DoS-bound: reject scripts that exceed {@link InputLimits.MAX_SCRIPT_BYTES}
   * BEFORE the interpreter starts. Each opcode iteration is cheap but a
   * multi-megabyte script would still pin the event loop; the largest
   * legitimate compiled Rúnar script (SLH-DSA-SHA2-256s, ~900 KB) fits
   * comfortably under the 1 MiB cap.
   */
  private assertScriptBytesUnderLimit(script: Uint8Array): void {
    if (script.length > InputLimits.MAX_SCRIPT_BYTES) {
      throw new CanonicalJsonError(
        'bytes',
        `ScriptVM: script length ${script.length} exceeds InputLimits.MAX_SCRIPT_BYTES (${InputLimits.MAX_SCRIPT_BYTES})`,
        { limit: InputLimits.MAX_SCRIPT_BYTES, actual: script.length },
      );
    }
  }

  // -------------------------------------------------------------------------
  // Public API — step mode
  // -------------------------------------------------------------------------

  /**
   * Load scripts for step-by-step execution.
   * Call `step()` repeatedly to advance one opcode at a time.
   */
  load(unlockingScript: Uint8Array, lockingScript: Uint8Array): void {
    this.assertScriptBytesUnderLimit(unlockingScript);
    this.assertScriptBytesUnderLimit(lockingScript);
    this.loadInternal(unlockingScript, lockingScript);
  }

  /**
   * Load scripts from hex strings for step-by-step execution.
   */
  loadHex(unlockingScriptHex: string, lockingScriptHex: string): void {
    this.load(
      unlockingScriptHex ? hexToBytes(unlockingScriptHex) : new Uint8Array(0),
      hexToBytes(lockingScriptHex),
    );
  }

  /**
   * Execute one opcode and return the step result.
   * Returns null once execution is complete.
   */
  step(): StepResult | null {
    const spend = this.spend;
    if (this._isComplete || !spend) return null;

    const pos = this.peekPosition();
    if (pos === null) {
      // Both scripts exhausted.
      this._isComplete = true;
      if (spend.ifStack.length > 0) {
        this._stepError = 'Unbalanced OP_IF/OP_ENDIF';
        this._isSuccess = false;
        return {
          offset: this.pc,
          opcode: 'END',
          mainStack: toBytes(spend.stack),
          altStack: toBytes(spend.altStack),
          error: this._stepError,
          context: spend.context === 'UnlockingScript' ? 'unlocking' : 'locking',
        };
      }
      this._isSuccess = this.topIsTruthy();
      return null;
    }

    const chunks = this.chunksFor(pos.context);
    const offsets = pos.context === 'unlocking' ? this.unlockingOffsets : this.lockingOffsets;
    const offset = offsets[pos.chunkIndex] ?? 0;
    const opcode = chunkOpcodeName(chunks[pos.chunkIndex]?.op ?? 0);

    try {
      this.countOp(chunks[pos.chunkIndex]?.op ?? 0);
      spend.step();
      this.afterStep();
      return {
        offset,
        opcode,
        mainStack: toBytes(spend.stack),
        altStack: toBytes(spend.altStack),
        context: pos.context,
      };
    } catch (e) {
      const msg = e instanceof ScriptError ? e.message : describeError(e);
      this._stepError = msg;
      this._isComplete = true;
      this._isSuccess = false;
      return {
        offset,
        opcode,
        mainStack: toBytes(spend.stack),
        altStack: toBytes(spend.altStack),
        error: msg,
        context: pos.context,
      };
    }
  }

  /** Current program counter (byte offset in the active script). */
  get pc(): number {
    const pos = this.peekPosition();
    if (pos === null) {
      const spend = this.spend;
      if (!spend) return 0;
      return spend.context === 'UnlockingScript' ? this.unlockingLength : this.lockingLength;
    }
    const offsets = pos.context === 'unlocking' ? this.unlockingOffsets : this.lockingOffsets;
    return offsets[pos.chunkIndex] ?? 0;
  }

  /** Which script is currently executing. */
  get context(): 'unlocking' | 'locking' {
    const pos = this.peekPosition();
    if (pos !== null) return pos.context;
    const spend = this.spend;
    if (!spend) return 'locking';
    return spend.context === 'UnlockingScript' ? 'unlocking' : 'locking';
  }

  /** Current main stack (copy). */
  get currentStack(): Uint8Array[] {
    return this.spend ? toBytes(this.spend.stack) : [];
  }

  /** Current alt stack (copy). */
  get currentAltStack(): Uint8Array[] {
    return this.spend ? toBytes(this.spend.altStack) : [];
  }

  /** Whether execution has completed (success or error). */
  get isComplete(): boolean {
    return this._isComplete;
  }

  /** Whether execution completed successfully (top of stack is truthy). */
  get isSuccess(): boolean {
    return this._isSuccess;
  }

  /**
   * Byte offset just after the most recently executed OP_CODESEPARATOR in the
   * currently running script (-1 if none executed yet). This is the scriptCode
   * start position for subsequent signature ops per post-Genesis BSV semantics.
   *
   * `Spend` tracks the separator as a CHUNK index (and uses
   * `chunks.slice(idx + 1)` as the subscript); this converts it to the byte
   * offset of that next chunk.
   */
  get lastCodeSeparator(): number {
    const spend = this.spend;
    if (!spend || spend.lastCodeSeparator === null) return -1;
    const inUnlocking = spend.context === 'UnlockingScript';
    const offsets = inUnlocking ? this.unlockingOffsets : this.lockingOffsets;
    const fallback = inUnlocking ? this.unlockingLength : this.lockingLength;
    return offsets[spend.lastCodeSeparator + 1] ?? fallback;
  }

  // -------------------------------------------------------------------------
  // Internal
  // -------------------------------------------------------------------------

  private loadInternal(unlockingScript: Uint8Array, lockingScript: Uint8Array): void {
    this.unlockingOffsets = chunkByteOffsets(unlockingScript);
    this.lockingOffsets = chunkByteOffsets(lockingScript);
    this.unlockingLength = unlockingScript.length;
    this.lockingLength = lockingScript.length;
    this._opsExecuted = 0;
    this._maxStackDepth = 0;
    this._isComplete = false;
    this._isSuccess = false;
    this._stepError = undefined;

    this.spend = new Spend({
      sourceTXID: SYNTHETIC_SOURCE_TXID,
      sourceOutputIndex: 0,
      sourceSatoshis: SYNTHETIC_SOURCE_SATOSHIS,
      lockingScript: new LockingScript([], lockingScript, undefined, false),
      transactionVersion: SYNTHETIC_TX_VERSION,
      otherInputs: [],
      outputs: [],
      unlockingScript: new UnlockingScript([], unlockingScript, undefined, false),
      inputIndex: 0,
      inputSequence: 0xffffffff,
      lockTime: 0,
      isRelaxed: this.flags.strictEncoding !== true,
    });
  }

  private chunksFor(context: 'unlocking' | 'locking'): { op: number; data?: number[] }[] {
    const spend = this.spend!;
    return (context === 'unlocking' ? spend.unlockingScript : spend.lockingScript).chunks;
  }

  /**
   * The chunk `Spend.step()` is about to execute, mirroring the context switch
   * in `step()`'s prologue. `null` once both scripts are exhausted.
   */
  private peekPosition(): Position | null {
    const spend = this.spend;
    if (!spend) return null;
    let inUnlocking = spend.context === 'UnlockingScript';
    let pc = spend.programCounter;
    if (inUnlocking && pc >= spend.unlockingScript.chunks.length) {
      inUnlocking = false;
      pc = 0;
    }
    const chunks = inUnlocking ? spend.unlockingScript.chunks : spend.lockingScript.chunks;
    if (pc >= chunks.length) return null;
    return { context: inUnlocking ? 'unlocking' : 'locking', chunkIndex: pc };
  }

  /** Consensus op counting: everything above OP_16 counts. */
  private countOp(op: number): void {
    if (op <= OP_16) return;
    this._opsExecuted++;
    if (this._opsExecuted > this.maxOps) {
      throw new ScriptError(`Operation limit exceeded (max ${this.maxOps})`);
    }
  }

  /** Harness stack bound + depth tracking, applied after each opcode. */
  private afterStep(): void {
    const spend = this.spend!;
    const depth = spend.stack.length + spend.altStack.length;
    if (depth > this._maxStackDepth) this._maxStackDepth = depth;
    if (depth > this.maxStackSize) {
      throw new ScriptError('Stack size limit exceeded');
    }
  }

  private topIsTruthy(): boolean {
    const spend = this.spend;
    if (!spend || spend.stack.length === 0) return false;
    return isTruthy(Uint8Array.from(spend.stack[spend.stack.length - 1]!));
  }

  /**
   * Drive `Spend` to completion. Mirrors `Spend.validate()`'s evaluation loop
   * (without its consensus wrappers — see the module header), adding the
   * harness op/stack bounds between opcodes.
   */
  private runToEnd(): VMResult {
    const spend = this.spend!;
    try {
      for (;;) {
        const pos = this.peekPosition();
        if (pos === null) break;
        const chunks = this.chunksFor(pos.context);
        this.countOp(chunks[pos.chunkIndex]?.op ?? 0);
        if (!spend.step()) break;
        this.afterStep();
      }
      if (spend.ifStack.length > 0) {
        return this.buildResult('Unbalanced OP_IF/OP_ENDIF');
      }
      this._isComplete = true;
      this._isSuccess = this.topIsTruthy();
      return this.buildResult();
    } catch (e) {
      const msg = e instanceof ScriptError ? e.message : describeError(e);
      this._isComplete = true;
      this._isSuccess = false;
      this._stepError = msg;
      return this.buildResult(msg);
    }
  }

  private buildResult(error?: string): VMResult {
    const spend = this.spend;
    const stack = spend ? toBytes(spend.stack) : [];
    const altStack = spend ? toBytes(spend.altStack) : [];
    const success = !error && stack.length > 0 && isTruthy(stack[stack.length - 1]!);
    return {
      success,
      stack,
      altStack,
      error,
      opsExecuted: this._opsExecuted,
      maxStackDepth: this._maxStackDepth,
    };
  }
}
